### PR TITLE
feat(v2): prefetching/ preloading API

### PR DESCRIPTION
**Pull request by endiliey**
## Motivation

Link prefetching is a browser mechanism, which utilizes browser idle time to download or prefetch documents that the user might visit in the near future. A web page provides a set of prefetching hints to the browser, and after the browser is finished loading the page, it begins silently prefetching specified documents and stores them in its cache. When the user visits one of the prefetched documents, it can be served up quickly out of the browser's cache.

### Description
Enable prefetching from anywhere. There is now some sort of api runner called `__docusaurus` in window that provides user some powerful utility function like prefetch. Prefetching is inspired from Gatsby. https://www.gatsbyjs.org/docs/how-code-splitting-works/
But we are not just limited to viewport, we allow prefetching from anywhere  !!

It is now possible in the future to make it such that:
- Doc page can prefetch next and previous pages
- Prefetch link if that link is on viewport (through Intersection Observer API) 
- etc

### Implementation

First of all, we need to understand that there are webpack **chunk names** and **chunk assets**
This is `chunk-map` manifest

```js
"component---theme-doc-body-f68": [ // webpack chunk name
    "/component---theme-doc-body-f68.9967d7798f71bc4d085e.css", // chunk assets
    "/component---theme-doc-body-f68.9967d7798f71bc4d085e.js" // chunk assets
  ],
```

In Docusaurus 2, every route is code-spliited, and there are chunk names associated with each route. 
For example: `component---theme-doc-body-f68`, `module---docs-commands-172-0af` and `metadata---docs-commandsa-60-e10` chunk names are related to route `/docs/commands`

General flow ?
From the routePath (for example: `/docs/commands`, we can infer all the related **webpack chunk names**. From that webpack chunk names, we can get all the **webpack chunk assets**. We then prefetch all those chunk assets. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

For example, **our home page now prefetch two url that people will most likely visit**. Making the site experience better.
https://github.com/facebook/Docusaurus/blob/163f706c87b724919aaf0be20e2cb20fcec3a443/website/pages/index.js#L65-L72

home page - prefetching

**Prefetch via console**
prefetch

### CODE REVIEW

**Comment by endiliey** [comment]
Deploy preview for *docusaurus-2* failed.

Built with commit 163f706c87b724919aaf0be20e2cb20fcec3a443

https://app.netlify.com/sites/docusaurus-2/deploys/5cb9bb9fd43a4e000833c014

File: packages/docusaurus/lib/client/App.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ function App() {
   const [context, setContext] = useState({});
   return (
     <DocusaurusContext.Provider
-      value={{siteConfig, ...metadata, ...context, setContext, routes}}>
```
Comment (endiliey) [inline packages/docusaurus/lib/client/App.js]: Not used in context. Should we include it ?

File: packages/docusaurus/lib/client/exports/ComponentCreator.js [hunk 2]
```diff
@@ -8,62 +8,67 @@
 import React from 'react';
 import Loadable from 'react-loadable';
 import Loading from '@theme/Loading';
-import routesAsyncModules from '@generated/routesAsyncModules';
+import routesChunkNames from '@generated/routesChunkNames';
 import registry from '@generated/registry';
 
 function ComponentCreator(path) {
-  const modules = routesAsyncModules[path];
+  /* chunkNames are mapping of the route module -> webpack chunk name
```
Comment (endiliey) [inline packages/docusaurus/lib/client/exports/ComponentCreator.js]: No logic changes here, just renaming and better comment. Ignore this file

File: packages/docusaurus/lib/client/exports/Link.js [hunk 3]
```diff
@@ -5,28 +5,23 @@
...
-  return !targetLink || isExternal ? (
+  const isInternal = internalRegex.test(targetLink);
+  return !targetLink || !isInternal ? (
     // eslint-disable-next-line jsx-a11y/anchor-has-content
     <a {...props} href={targetLink} />
   ) : (
     <Perimeter
       padding={preloadProximity}
-      onBreach={() => preload(routes, targetLink)}
+      onBreach={() => window.__docusaurus.preload(targetLink)}
```
Comment (endiliey) [inline packages/docusaurus/lib/client/exports/Link.js]: Use centralized docusaurus preload method because we want to dedupe prefetch and preload

File: packages/docusaurus/lib/client/prefetch.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const support = function(feature) {
```
Comment (endiliey) [inline packages/docusaurus/lib/client/prefetch.js]: Taken from gatsby prefetch function 😉.

File: packages/docusaurus/lib/commands/build.js [hunk 5]
```diff
@@ -56,6 +57,10 @@ module.exports = async function build(siteDir, cliOptions = {}) {
       new CleanWebpackPlugin({verbose: false}),
       // Visualize size of webpack output files with an interactive zoomable treemap.
       cliOptions.bundleAnalyzer && new BundleAnalyzerPlugin(),
+      // Generate client manifests file that will be used for server bundle
```
Comment (endiliey) [inline packages/docusaurus/lib/commands/build.js]: Only generate this manifest in production.,

File: packages/docusaurus/lib/client/templates/index.html.template.ejs [hunk 6]
```diff
@@ -7,6 +7,5 @@
   </head>
   <body>
     <div id="__docusaurus"></div>
-    <script src="<%= htmlWebpackPlugin.files.chunks.main.entry %>"></script>
```
Comment (endiliey) [inline packages/docusaurus/lib/client/templates/index.html.template.ejs]: Useless because html webpack plugin auto inject anyway

File: packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
```
Comment (endiliey) [inline packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js]: This is the main logic to build chunk names -> chunk assets mapping. This plugin is not available anywhere ! Not a fork !

File: packages/docusaurus/package.json [hunk 8]
```diff
@@ -51,7 +51,7 @@
     "front-matter": "^3.0.1",
     "fs-extra": "^7.0.0",
     "globby": "^9.1.0",
-    "html-webpack-plugin": "^3.2.0",
+    "html-webpack-plugin": "^4.0.0-beta.5",
```
Comment (endiliey) [inline packages/docusaurus/package.json]: upgrade for better

File: website/pages/index.js [hunk 9]
```diff
@@ -61,6 +62,15 @@ function Home() {
   const context = useContext(DocusaurusContext);
   const {siteConfig = {}} = context;
 
+  const feedbackUrl = `${siteConfig.baseUrl}feedback/`;
+  const gettingStartedUrl = `${siteConfig.baseUrl}docs/installation`;
+
+  useEffect(() => {
```
Comment (endiliey) [inline website/pages/index.js]: Prefetch two most likely visited pages.

File: packages/docusaurus/lib/client/App.js (hunk 1 above)
Comment (yangshun) [reply packages/docusaurus/lib/client/App.js]: What is not used?

File: packages/docusaurus/lib/client/docusaurus.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+const docusaurus = {
+  prefetch: routePath => {
+    if (!canPrefetchOrLoad(routePath)) {
+      return false;
+    }
+    fetched[routePath] = true;
+
+    // Find all chunk names needed for that particular route
+    const chunkNamesNeeded = [];
+    function dfs(target) {
```
Comment (yangshun) [inline packages/docusaurus/lib/client/docusaurus.js]: Does this make "visited" chunks get added twice? Not too confident of the algorithm here, not very conventional DFS 😝

File: packages/docusaurus/lib/client/exports/ComponentCreator.js [hunk 11]
```diff
@@ -8,62 +8,67 @@
...
+  Example:
+    {
+      "component": "component---theme-pages-767",
+      "content": {
+        "1": "module---06-c-47f",
```
Comment (yangshun) [inline packages/docusaurus/lib/client/exports/ComponentCreator.js]: Why is this changed to an object? It should remain as an array. Objects have no ordering but arrays do and over here the ordering matters.

File: packages/docusaurus/lib/client/prefetch.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    const parentElement =
+      document.getElementsByTagName(`head`)[0] ||
+      document.getElementsByName(`script`)[0].parentNode;
+    parentElement.appendChild(link);
+  });
+};
+
+const xhrPrefetchStrategy = function(url) {
+  return new Promise((resolve, reject) => {
+    const req = new XMLHttpRequest();
```
Comment (yangshun) [inline packages/docusaurus/lib/client/prefetch.js]: Use `fetch` instead

File: packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js (hunk 7 above)
Comment (yangshun) [reply packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js]: Should this be a package on its own?

File: website/pages/index.js [hunk 13]
```diff
@@ -61,6 +62,15 @@ function Home() {
   const context = useContext(DocusaurusContext);
   const {siteConfig = {}} = context;
 
+  const feedbackUrl = `${siteConfig.baseUrl}feedback/`;
```
Comment (yangshun) [inline website/pages/index.js]: We need a API (wrapper function) for these and not let people concatenating the URLs themselves.

File: website/pages/index.js [hunk 14]
```diff
@@ -61,6 +62,15 @@ function Home() {
...
+  const gettingStartedUrl = `${siteConfig.baseUrl}docs/installation`;
+
+  useEffect(() => {
+    // Prefetch feedback pages & getting started pages
+    __docusaurus.prefetch(feedbackUrl) &&
```
Comment (yangshun) [inline website/pages/index.js]: I get why we write it like this, but I think we should put them on separate lines. The difference is not very significant.

File: packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js (hunk 7 above)
Comment (endiliey) [reply packages/docusaurus/lib/webpack/plugins/ChunkManifestPlugin.js]: This is more fit to use for docusaurus itself actually. Kinda hard to maintain it as a package and had to make it general

File: packages/docusaurus/lib/client/exports/ComponentCreator.js (hunk 11 above)
Comment (endiliey) [reply packages/docusaurus/lib/client/exports/ComponentCreator.js]: Actually there is no logic change here. I just renamed `routesAsyncModules` to `routesChunkName` and put the example structure here.

```js
/* chunkNames are mapping of the route module -> webpack chunk name
  Example:
    {
      "component": "component---theme-pages-767",
      "content": {
        "1": "module---06-c-47f",
      }
    }
  */
```

To describe that later on it will be converted to
```js
/* Prepare opts data that react-loadable needs
  https://github.com/jamiebuilds/react-loadable#declaring-which-modules-are-being-loaded
  Example: (with above example as input)
  - optsLoader: 
  {
    component: () => import('./Pages.js'),
    content.1: () => import('./doc1.md'),
  }
  - optsModules: ['./Pages.js', './doc1.md']
  - optsWebpack: [require.resolveWeak('./Pages.js'), require.resolveWeak('./doc1.md')]
  */
```

You are the one who made the structure like this 😃 https://github.com/facebook/Docusaurus/pull/1366/files

File: packages/docusaurus/lib/client/prefetch.js (hunk 12 above)
Comment (endiliey) [reply packages/docusaurus/lib/client/prefetch.js]: Actually this code is copied from Gatsby https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/cache-dir/prefetch.js

I think it's battle tested 😆

File: website/pages/index.js (hunk 13 above)
Comment (endiliey) [reply website/pages/index.js]: Let's add it on hack md. And think of the wrapper function name.

This PR is already quite big

File: packages/docusaurus/lib/client/docusaurus.js (hunk 10 above)
Comment (endiliey) [reply packages/docusaurus/lib/client/docusaurus.js]: Hahah this algorithm need to be revised. I'll reimplement ha ha ha

File: packages/docusaurus/lib/client/clientEntry.js [hunk 15]
```diff
@@ -11,10 +11,12 @@ import {BrowserRouter} from 'react-router-dom';
 
 import App from './App';
 import preload from './preload';
+import docusaurus from './docusaurus';
 import routes from '@generated/routes'; // eslint-disable-line
 
 // Client-side render (e.g: running in browser) to become single-page application (SPA).
 if (typeof window !== 'undefined' && typeof document !== 'undefined') {
+  window.__docusaurus = docusaurus;
```
[the remark this answers is not in the document]
Comment (endiliey) [reply packages/docusaurus/lib/client/clientEntry.js]: There will be a cyclic dependency if downstream try to import it. Mainly due to `routes`

It is also easier to do `__docusaurus` instead of `import + then call the function`.
This make it possible to have one source of truth
Algolia is doing this as well, the api runner is actually in `window.__algolia`
Gatsby prefetching is done like this too. `window.__loader`

Their prefetch is
```js
___loader.enqueue(parsePath(this.props.to).pathname)
```

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-link/src/index.js#L61

**Comment by endiliey** [comment]
Bundle increase is around ~2kbs. Gzipped is probably ard ~0.60kb difference.

If we could delete blogMetadata.json, docsMetadata.json and pagesMetadata.json, we can save a lot. Maybe ~15kb

Before
image


After
image

File: packages/docusaurus/lib/client/exports/ComponentCreator.js (hunk 11 above)
Comment (yangshun) [reply packages/docusaurus/lib/client/exports/ComponentCreator.js]: I guess it's not wrong to put that as an example, but the key `1` is misleading. I thought you changed my array to keys 😓

File: website/pages/index.js [hunk 16]
```diff
@@ -61,6 +62,16 @@ function Home() {
...
   const {siteConfig = {}} = context;
 
+  // TODO: (wrapper function) API so that user won't need to concatenate url manually
+  const feedbackUrl = `${siteConfig.baseUrl}feedback/`;
+  const gettingStartedUrl = `${siteConfig.baseUrl}docs/installation`;
+
+  useEffect(() => {
+    // Prefetch feedback pages & getting started pages
+    __docusaurus.prefetch(feedbackUrl);
+    __docusaurus.prefetch(gettingStartedUrl);
```
Comment (radekmie) [inline website/pages/index.js]: As I've tested the latest `master`, it _sometimes_ is not the object it expects. The root node - the one with `id="__docusaurus"` - is a _named element_, therefore it's a global. If the code which sets `window.__docusaurus` runs _after_ this very line, an error is thrown here.

Comment (endiliey) [reply website/pages/index.js]: Thanks for reporting @radekmie. It is very interesting to see someone checking out this PR. We will most likely rename since there is a conflict
